### PR TITLE
[SymbolGraph] Add information about "inherited docs" for synthesized symbols

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
Resolves rdar://75741632

This PR adds handling to symbol-graph related code about the concept of "inherited docs", on synthesized symbols inherited from default implementations and classes. There are two primary updates as part of this PR:

1. For "synthesized" symbols, there is a new field added to the relationship tying the synthesized symbol to its parent: `"sourceOrigin"`. This contains the USR and the display path of the parent symbol.
2. There is a new flag added to the frontend/driver/symbol-graph tool: `-skip-inherited-docs`. This changes the symbol-graph code to skip emitting the `"docComment"` field on synthesized symbols. Since the doc comment is going to be the same as the parent symbol's, this allows a user or tool to skip emitting duplicate information, optimizing the size of the output symbol graph.

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
@swift-ci Please test

**Comment by QuietMisdreavus** [comment]
@swift-ci Please build toolchain macOS platform

**Comment by QuietMisdreavus** [comment]
After some more discussion, it looks like we should handle docs inheritance for symbols that aren't synthesized. For example, in my test, the extension's default implementation appears as its own symbol, but inherits its documentation from the protocol. I've pushed a new commit that extends the "inherited docs" behavior to these non-synthesized symbols as well.

@swift-ci Please test

**Comment by QuietMisdreavus** [comment]
@swift-ci Please build toolchain macOS platform

**Comment by QuietMisdreavus** [comment]
While testing the toolchain, we ran into some issues where synthesized symbols were directly inheriting docs from the "base" symbol, and losing the `sourceOrigin` field because the code assumed it wasn't inheriting docs. I've pushed a new commit that adds this field back to synthesized symbols that have docs to inherit.

@swift-ci Please test

**Comment by QuietMisdreavus** [comment]
@swift-ci Please build toolchain macOS platform

File: lib/SymbolGraphGen/Edge.cpp [hunk 1]
```diff
@@ -57,5 +57,27 @@ void Edge::serialize(llvm::json::OStream &OS) const {
...
       }
     }
+    
+    const ValueDecl *InheritingDecl = nullptr;
+    if (const auto *ID = Source.getDeclInheritingDocs()) {
+      if (Target.getSymbolDecl() == ID || Source.getSynthesizedBaseTypeDecl())
+        InheritingDecl = ID;
+    }
+    
+    // If our source symbol is a synthesized decl, write in information about
```
Comment (franklinsch) [inline lib/SymbolGraphGen/Edge.cpp:67]:
```suggestion
    // If our source symbol is a inheriting decl, write in information about
```
?

**Review by franklinsch — APPROVED** [review]
LGTM

File: lib/SymbolGraphGen/Edge.cpp (hunk 1 above)
Comment (QuietMisdreavus) [reply lib/SymbolGraphGen/Edge.cpp:67]: Good catch!

**Comment by QuietMisdreavus** [comment]
@swift-ci Please smoke test

**Pull request merged**